src/lazy/encoder/write_as_ion.rs

```diff
@@ -45,229 +41,210 @@ impl<T> WriteAsIon for T
         writer.write_string(self)
     }
 }
 
 impl<'b> WriteAsIonValue for &'b [u8] {
-    fn write_as_ion_value<'a, W: Write + 'a, E: LazyEncoder<W>, V: ValueWriter<'a, W, E>>(
-        &self,
-        writer: V,
-    ) -> IonResult<()> {
+    fn write_as_ion_value<V: ValueWriter>(&self, writer: V) -> IonResult<()> {
         writer.write_blob(self)
     }
 }
 
 impl<const N: usize> WriteAsIonValue for [u8; N] {
-    fn write_as_ion_value<'a, W: Write + 'a, E: LazyEncoder<W>, V: ValueWriter<'a, W, E>>(
-        &self,
-        writer: V,
-    ) -> IonResult<()> {
+    fn write_as_ion_value<V: ValueWriter>(&self, writer: V) -> IonResult<()> {
         writer.write_blob(self)
     }
 }
 
 impl WriteAsIonValue for Blob {
-    fn write_as_ion_value<'a, W: Write + 'a, E: LazyEncoder<W>, V: ValueWriter<'a, W, E>>(
-        &self,
-        writer: V,
-    ) -> IonResult<()> {
+    fn write_as_ion_value<V: ValueWriter>(&self, writer: V) -> IonResult<()> {
         writer.write_blob(self)
     }
 }
 
 impl WriteAsIonValue for Clob {
-    fn write_as_ion_value<'a, W: Write + 'a, E: LazyEncoder<W>, V: ValueWriter<'a, W, E>>(
-        &self,
-        writer: V,
-    ) -> IonResult<()> {
+    fn write_as_ion_value<V: ValueWriter>(&self, writer: V) -> IonResult<()> {
         writer.write_clob(self)
     }
 }
 
-impl<T> WriteAsIonValue for &T
-where
-    T: WriteAsIonValue,
-{
-    fn write_as_ion_value<'a, W: Write + 'a, E: LazyEncoder<W>, V: ValueWriter<'a, W, E>>(
-        &self,
-        writer: V,
-    ) -> IonResult<()> {
+impl<T: WriteAsIonValue> WriteAsIonValue for &T {
+    fn write_as_ion_value<V: ValueWriter>(&self, writer: V) -> IonResult<()> {
         (*self).write_as_ion_value(writer)
     }
 }
 
-impl<T> WriteAsIonValue for &[T]
-where
-    for<'x> &'x T: WriteAsIon,
-{
-    fn write_as_ion_value<'a, W: Write + 'a, E: LazyEncoder<W>, V: ValueWriter<'a, W, E>>(
-        &self,
-        writer: V,
-    ) -> IonResult<()> {
-        let mut list = writer.list_writer()?;
-        for value in *self {
-            list.write(value)?;
+macro_rules! impl_write_as_ion_value_for_iterable {
+    ($iterable:ty, $item:ident $(, const $n:ident: $n_type:ty)?) => {
+        impl<$item $(, const $n: $n_type)?> WriteAsIonValue for $iterable
+        where
+            for<'a> &'a $item: WriteAsIon,
+        {
+            fn write_as_ion_value<V: ValueWriter>(&self, writer: V) -> IonResult<()> {
+                writer.write_list(|list| {
+                    for value in self.iter() {
+                        list.write(value)?;
+                    }
+                    Ok(())
+                })
+            }
         }
-        list.end()?;
-        Ok(())
-    }
+    };
 }
 
-impl<T, const N: usize> WriteAsIonValue for [T; N]
+impl_write_as_ion_value_for_iterable!(Vec<T>, T);
+impl_write_as_ion_value_for_iterable!(&[T], T);
+impl_write_as_ion_value_for_iterable!([T; N], T, const N: usize);
+
+pub trait WriteAsSExp<T>: Sized
```

Comment:
I assume this is needed so that you can indicate that an iterable of some sort should be written as a sexp instead of a list (which I assume is the default).